src/utils.jl

```diff
@@ -0,0 +1,82 @@
+"""
+    indices_of_unique(x) -> Dict
+
+Return a `Dict` whose keys are the unique elements of `x` and whose values are the
+corresponding indices in `x`.
+"""
+function indices_of_unique(x)
+    d = Dict{eltype(x),Vector{Int}}()
+    for (i, xi) in enumerate(x)
+        if haskey(d, xi)
+            push!(d[xi], i)
+        else
+            d[xi] = [i]
+        end
```

Comment:
This can be made more efficient by not looking up the key twice. One could e.g. use
```suggestion
        d_xi = get!(d, xi) do
            return Int[]
        end
        push!(d_xi, i)
```

Apart from that, it seems like a function that could exist e.g. in StatsBase (similar to `proportionmap` etc.). Did you check that?

Reply:
Sounds good, I just want to make sure we use existing functionality. If it doesn't exist yet that's unfortunate but, of course, then we should use our own implementation.

Reply:
Actually, it seems `indicatormap` returns the information we are interested in: https://juliastats.org/StatsBase.jl/stable/misc/#StatsBase.indicatormat But maybe it's not the desired output format for our purposes.

Reply:
It's similar, yes, but a little clunky. e.g. here's how we could get the vector of indices:
```julia
using SparseArrays
map(first ∘ findnz ∘ sparse, eachslice(indicatormat(x; sparse=true); dims=1))
```
But I still think it makes more sense to try to upstream the functionality we want, since often something like what we want will be more convenient for the user.

Reply:
Yes, I agree. That seems a bit inconvenient.